OpenRA.Mods.Common/Widgets/Selection.cs

```diff
@@ -0,0 +1,245 @@
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Graphics;
+using OpenRA.Primitives;
+
+namespace OpenRA.Mods.Common.Widgets
+{
+	class Selection
+	{
+		public enum States { Empty, Ready, Active, Inactive }
+		public States State = States.Empty;
+		public LabelWithSelectionWidget SelectionWidget = null;
+		public int2 StartingLocation;
+		public int2 EndingLocation;
+
+		RenderedText TextAsDrawn
+		{
+			get
+			{
+				return RenderedText.From(
+						SelectionWidget.GetTextContent(),
+						SelectionWidget.GetFont(),
+						SelectionWidget.GetPosition());
+			}
+		}
+
+		public int Start
+		{
+			get
+			{
+				return Math.Min(
+						TextAsDrawn.ClosestCharacter(StartingLocation).Index,
+						TextAsDrawn.ClosestCharacter(EndingLocation).Index);
+			}
+		}
+
+		public int End
+		{
+			get
+			{
+				return Math.Max(
+						TextAsDrawn.ClosestCharacter(StartingLocation).Index,
+						TextAsDrawn.ClosestCharacter(EndingLocation).Index);
+			}
+		}
+
+		public Selection() { }
+
+		public bool OwnedBy(LabelWidget widget)
+		{
+			return SelectionWidget == widget;
+		}
+
+		public bool HandleMouseDown(LabelWithSelectionWidget widget, int2 location)
+		{
+			if (object.ReferenceEquals(widget, null))
+				return false;
+
+			switch (State)
+			{
+				case States.Empty:
+				case States.Inactive:
+					SelectionWidget = widget;
+					StartingLocation = location;
+					EndingLocation = location;
+					State = States.Ready;
+					return true;
+				default:
+					return false;
+			}
+		}
+
+		public bool HandleMouseMove(int2 location)
+		{
+			if (object.ReferenceEquals(SelectionWidget, null))
+				return false;
+
+			switch (State)
+			{
+				case States.Ready:
+					EndingLocation = location;
+					State = States.Active;
+					return true;
+				case States.Active:
+					EndingLocation = location;
+					return true;
+				default:
+					return false;
+			}
+		}
+
+		public bool HandleMouseUp()
+		{
+			switch (State)
+			{
+				case States.Active:
+					State = States.Inactive;
+					return true;
+				case States.Ready:
+					State = States.Empty;
+					SelectionWidget = null;
+					return true;
+				default:
+					return false;
+			}
+		}
+
+		public bool HandleLooseMouseFocus()
+		{
+			State = States.Empty;
+			SelectionWidget = null;
+			return true;
+		}
+
+		class RenderedText
+		{
+			public class Character
+			{
+				// NOTE: What happens when we handle unicode?
+				public char Value;
+				public int Index;
+				public int LineIndex;
+				public Rectangle Bounds;
+
+				public static Character From(char val, int index, int lineIndex, SpriteFont font, int2 topLeft)
+				{
+					var bounds = Rectangle.FromCorners(topLeft, topLeft + font.Measure(val.ToString()));
```

Comment:
I'd prefer it if we could define `var bottomRight = topLeft + font.Measure(val.ToString())` and then use `var bounds = Rectangle.FromLTRB(topLeft.X, topLeft.Y, bottomRight.X, bottomRight.Y)`. Adding lots of similar-but-different ways to create objects can cause real headaches when refactoring code, and I don't think the 3 uses in this file really justify adding a new method.